File: LinkedList/03.LinkedListCycle2/LinkedListCycle2.py
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: return None

        try:
            slow = head
            fast = head.next
            while slow is not fast:
                slow = slow.next
                fast = fast.next.next
        except:
            return None

        slow = slow.next
        helpNode = head
        while slow is not helpNode:
            slow = slow.next
            helpNode = helpNode.next

        return slow

File: LinkedList/LinkedListCycle2/LinkedListCycle2.cpp
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode *fast = head, *slow = head;
        while (fast && fast->next) {
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow) {
                fast = head;
                while (fast != slow) {
                   fast = fast->next;
                   slow = slow->next;
                }
                return fast;
            }
        }
        return nullptr;
    }
};